### PR TITLE
Oauth2 Support

### PR DESCRIPTION
**Pull request by danielpetisme**

### CODE REVIEW

**Comment by danielpetisme** [comment]
I think the PR will be quite big.
Friends don't let friends wait until the end of the PR to have a full style review 😉

**Comment by avdev4j** [comment]
I'm sorry you will need to update your branch :D. But probably better to do it now :).

I also put the PR as draft to avoid merge it accidentally

**Comment by danielpetisme** [comment]
@avdev4j @joewhite101 @mraible It has been a long fight but Oauth2 finally works with Quarkus!!

Please if you can do a review/test session, it would definitively help me.

File: generators/common/templates/quarkus/README.md.ejs [hunk 1]
```diff
@@ -104,19 +104,13 @@ docker-compose -f src/main/docker/keycloak.yml up
...
-        registration:
-          oidc:
-            client-id: web_app
-            client-secret: web_app
+```properties
+quarkus.oidc.enabled=true
+quarkus.oidc.auth-server-url=http://localhost:9080/auth/realms/jhipster/
+quarkus.oidc.client-id=web_app
+quarkus.oidc.credentials.secret=web_app
+quarkus.oidc.authentication.scopes=profile
```
Comment (mraible) [inline generators/common/templates/quarkus/README.md.ejs:112]: Does quarkus send `openid` by default? This usually is something like `openid,profile,email` or it's space-delimited.

Comment (mraible) [reply generators/common/templates/quarkus/README.md.ejs:112]: Related: we just use the defaults for Spring Security, which I believe are `openid email profile`, but I'm not certain.

File: generators/server/files.js [hunk 2]
```diff
@@ -61,12 +61,15 @@ const serverFiles = {
...
+            path: SERVER_MAIN_RES_DIR,
+            templates: [{ file: 'default_banner.txt', method: 'copy', noEjs: true }, 'application.properties']
+        },
+        {
+            condition: generator => generator.authenticationType === 'jwt',
             path: SERVER_MAIN_RES_DIR,
             templates: [
-                { file: 'default_banner.txt', method: 'copy', noEjs: true },
                 { file: 'jwt/privateKey.pem', method: 'copy', noEjs: true },
                 { file: 'META-INF/resources/publicKey.pem', method: 'copy', noEjs: true },
```
Comment (mraible) [inline generators/server/files.js:72]: We don't need public/private keys for Spring Security. Are you sure it's necessary here? [This blog post shows Quarkus + OIDC with Okta](https://developer.okta.com/blog/2020/04/08/kafka-streams) and I don't see any keys (except to secure Kafka).

File: generators/server/files.js [hunk 3]
```diff
@@ -82,6 +85,11 @@ const serverFiles = {
         {
             path: SERVER_TEST_RES_DIR,
             templates: ['application.properties']
+        },
+        {
+            condition: generator => generator.authenticationType === 'oauth2',
+            path: SERVER_TEST_RES_DIR,
+            templates: [{ file: 'jhipster-test-realm.json', method: 'copy', noEjs: true }]
```
Comment (mraible) [inline generators/server/files.js:92]: How is this different from the regular one? We only have one for Spring Security.

File: generators/server/files.js [hunk 4]
```diff
@@ -99,14 +107,24 @@ const serverFiles = {
...
                     renameTo: generator => `${generator.javaDir}/TestResources.java`
                 }
             ]
+        },
+        {
+            condition: generator => generator.authenticationType === 'oauth2',
+            path: SERVER_TEST_SRC_DIR,
+            templates: [
+                {
+                    file: 'package/KeycloakServerTestResource.java',
```
Comment (mraible) [inline generators/server/files.js:124]: We shouldn't have any classes with an identity provider in its name because it should be easy to switch from one to another without changing any code. For Spring Security, we mock the provider in a [`TestSecurityConfiguration`](https://github.com/jhipster/generator-jhipster/blob/main/generators/server/templates/src/test/java/package/config/TestSecurityConfiguration.java.ejs) class.

File: generators/server/templates/quarkus/src/main/java/package/service/UserService.java.ejs [hunk 5]
```diff
@@ -317,6 +340,110 @@ public class UserService {
...
+    /**
+     * Returns the user from an OAuth 2.0 login or resource server with JWT.
+     * Synchronizes the user in the local repository.
+     *
+     * @param jwt the authentication token.
+     * @return the user from the authentication.
+     */
+     public UserDTO getUserFromAuthentication(JsonWebToken jwt) {
+        User user = getUser(jwt);
+//        var authorities = jwt.claim()
```
Comment (mraible) [inline generators/server/templates/quarkus/src/main/java/package/service/UserService.java.ejs:392]: Remove this line if it's not needed.

File: generators/server/templates/quarkus/src/main/java/package/service/UserService.java.ejs [hunk 6]
```diff
@@ -317,6 +340,110 @@ public class UserService {
...
+
+    private static User getUser(JsonWebToken jwt) {
+        User user = new User();
+        // handle resource server JWT, where sub claim is email and uid is ID
+        if (jwt.getClaim("uid") != null) {
+            user.id = 1L; // (String) jwt.getClaim("uid");
+            user.login = jwt.getClaim("sub");
+        } else {
+            user.id = 1L; // (String) jwt.getClaim("sub");
+        }
```
Comment (mraible) [inline generators/server/templates/quarkus/src/main/java/package/service/UserService.java.ejs:404]: Does this still need some work? It seems strange to set the ID to 1.

File: generators/server/templates/quarkus/src/main/java/package/service/UserService.java.ejs [hunk 7]
```diff
@@ -317,6 +340,110 @@ public class UserService {
...
+        if (jwt.getClaim("preferred_username") != null) {
+            user.login = ((String) jwt.getClaim("preferred_username")).toLowerCase();
+        } else if (user.login == null) {
+            user.login = "111"; //user.id;
```
Comment (mraible) [inline generators/server/templates/quarkus/src/main/java/package/service/UserService.java.ejs:408]: Seems strange.

File: generators/server/templates/quarkus/src/main/java/package/web/rest/UserOauth2Controller.java.ejs [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+        return;
+    }
+
+    @GET
+    @Path("login/oauth2/code/oidc")
+    @Authenticated
+    public Response callback(@Context UriInfo uriInfo) {
+        return Response.status(Response.Status.FOUND).location(uriInfo.getBaseUri()).build();
+    }
+}
```
Comment (mraible) [inline generators/server/templates/quarkus/src/main/java/package/web/rest/UserOauth2Controller.java.ejs:58]: Is this class necessary so the redirect URIs match the default for JHipster?

File: generators/server/templates/quarkus/src/main/resources/application.properties.ejs [hunk 9]
```diff
@@ -36,6 +36,7 @@ quarkus.micrometer.export.prometheus.path=/management/jhimetrics
 
 quarkus.smallrye-health.root-path=/management/health
 
+<%if (authenticationType === 'jwt') { _%>
 mp.jwt.verify.publickey.location=META-INF/resources/publicKey.pem
```
Comment (mraible) [inline generators/server/templates/quarkus/src/main/resources/application.properties.ejs:40]: Do you have to create public/private keys for JWTs? If so, are these generated for each new app, or are they the same one for each app?

File: generators/server/templates/quarkus/src/main/resources/application.properties.ejs [hunk 10]
```diff
@@ -74,3 +76,18 @@ quarkus.container-image.group=
 quarkus.container-image.name=<%= baseName %>
 quarkus.container-image.tag=latest
 <%_ } _%>
+
+<%if (authenticationType === 'oauth2') { _%>
+# Oauth 2.0 and OIDC
```
Comment (mraible) [inline generators/server/templates/quarkus/src/main/resources/application.properties.ejs:81]:
```suggestion
# OAuth 2.0 and OIDC
```

File: generators/server/templates/quarkus/src/main/resources/application.properties.ejs [hunk 11]
```diff
@@ -74,3 +76,18 @@ quarkus.container-image.group=
...
+quarkus.oidc.enabled=true
+quarkus.oidc.auth-server-url=http://localhost:9080/auth/realms/jhipster/
+quarkus.oidc.client-id=web_app
+quarkus.oidc.credentials.secret=web_app
+quarkus.oidc.application-type=hybrid
+quarkus.oidc.authentication.scopes=profile,address,email,jhipster,microprofile-jwt,offline_access,phone
```
Comment (mraible) [inline generators/server/templates/quarkus/src/main/resources/application.properties.ejs:87]: We should only use standard OIDC scopes here so users don't have to set up new scopes on their identity provider. https://openid.net/specs/openid-connect-core-1_0.html#ScopeClaims

In other words, remove `jhipster` and `microprofile-jwt` (unless they're necessary for some reason).

File: generators/server/templates/quarkus/src/test/java/package/KeycloakServerTestResource.java.ejs [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+import org.testcontainers.containers.GenericContainer;
+import org.testcontainers.containers.wait.strategy.Wait;
+
+import java.util.Collections;
+import java.util.Map;
+
+public class KeycloakServerTestResource implements QuarkusTestResourceLifecycleManager {
+
+    private final Logger log = LoggerFactory.getLogger(KeycloakServerTestResource.class);
+
```
Comment (mraible) [inline generators/server/templates/quarkus/src/test/java/package/KeycloakServerTestResource.java.ejs:34]: Remove extra line break.

File: generators/server/templates/quarkus/src/test/java/package/KeycloakServerTestResource.java.ejs [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+
+    public GenericContainer keycloak;
+
+    @Override
+    public Map<String, String> start() {
+        try {
+            keycloak = new KeycloakContainer("jboss/keycloak:latest");
+            keycloak.start();
+            System.out.println("[INFO] Keycloack server started in TCP server mode");
```
Comment (mraible) [inline generators/server/templates/quarkus/src/test/java/package/KeycloakServerTestResource.java.ejs:43]:
```suggestion
            System.out.println("[INFO] Keycloak server started in TCP server mode");
```

File: generators/server/templates/quarkus/src/test/java/package/TestResources.java.ejs [hunk 14]
```diff
@@ -21,5 +21,10 @@ package <%=packageName%>;
 import io.quarkus.test.common.QuarkusTestResource;
 import io.quarkus.test.h2.H2DatabaseTestResource;
 
+<%_ if (databaseType !== 'no') { _%>
 @QuarkusTestResource(H2DatabaseTestResource.class)
+<%_ } _%>
+<%_ if (authenticationType === 'oauth2') { _%>
+@QuarkusTestResource(KeycloakServerTestResource.class)
```
Comment (mraible) [inline generators/server/templates/quarkus/src/test/java/package/TestResources.java.ejs:28]: It'd be cool if we could mock the identity provider so we don't need to spin up a container (like we can with Spring Security).

File: generators/server/templates/quarkus/src/test/resources/application.properties.ejs [hunk 15]
```diff
@@ -59,10 +63,25 @@ quarkus.swagger-ui.path=/swagger-ui
...
 <%if (buildTool === 'gradle') { _%>
+
 # Docker image build properties, for more information https://quarkus.io/guides/container-image
 quarkus.container-image.group=
 quarkus.container-image.name=<%= baseName %>
 quarkus.container-image.tag=latest
 <%_ } _%>
+
+<%if (authenticationType === 'oauth2') { _%>
+# Oauth 2.0 and OIDC
```
Comment (mraible) [inline generators/server/templates/quarkus/src/test/resources/application.properties.ejs:75]:
```suggestion
# OAuth 2.0 and OIDC
```

File: generators/server/templates/quarkus/src/test/resources/application.properties.ejs [hunk 16]
```diff
@@ -59,10 +63,25 @@ quarkus.swagger-ui.path=/swagger-ui
...
+quarkus.oidc.auth-server-url=http://localhost:8180/auth/realms/jhipster-test/
+quarkus.oidc.client-id=web_app
+quarkus.oidc.credentials.secret=web_app
+quarkus.oidc.application-type=service
+quarkus.oidc.authentication.scopes=profile,address,email,jhipster,microprofile-jwt,offline_access,phone
+quarkus.oidc.authentication.cookie-path=/
+quarkus.oidc.authentication.redirect-path=/login/oauth2/code/oidc
+quarkus.oidc.authentication.restore-path-after-redirect=false
+
+jhipster.oidc.logout-url=http://localhost:8180/auth/realms/jhipster-test/protocol/openid-connect/logout
```
Comment (mraible) [inline generators/server/templates/quarkus/src/test/resources/application.properties.ejs:86]: It'd be nice if these values were fake and not tied to a particular IdP (identity provider).

**Review by mraible — COMMENTED** [review]
I gave this PR a quick glance but did not try generating an app and making sure it works. If you enable e2e tests on a generated Quarkus app in CI, that should be "good enough". If it's in a JDL, even better because we can test it easier locally.

File: generators/server/files.js (hunk 2 above)
Comment (danielpetisme) [reply generators/server/files.js:72]: Are you sure you're looking the version of this file?
https://github.com/jhipster/generator-jhipster-quarkus/blob/d9bd816508f607d5c0b838821d94de8ad5d45fa5/generators/server/files.js#L68
I've added a condition to generate the keys only when jwt is choosen

**Comment by danielpetisme** [comment]
@mraible Quarkus does not provide a way to have custom security context just as Spring does.
For testing purpose, I easier rely on keycloak or have to mock the IdP by simulating oauth2 http calls

**Comment by danielpetisme** [comment]
@mraible I mocked the OIDC server 🎉 
This permits to prevent a container to spin up + be independent of the OIDC provider.
Please have a look to [MockOidcServerTestResource](https://github.com/jhipster/generator-jhipster-quarkus/blob/a586b0a28e5a923a51be606c393b72b9e1fb232a/generators/server/templates/quarkus/src/test/java/package/MockOidcServerTestResource.java.ejs).

**Comment by danielpetisme** [comment]
Regarding the logout url, I do not actually test the logout but the `api/logout` json response. This means we do not need a meaningful url.
To prevent misunderstanding I make it explicit I'm expecting a dummy value. See 8218d7005b38dc9bf3d58b8d0d021450c32ec1f2

File: generators/server/files.js (hunk 3 above)
Comment (danielpetisme) [reply generators/server/files.js]: Removed via the Mock Oidc Test Resource

File: generators/server/templates/quarkus/src/test/java/package/TestResources.java.ejs (hunk 14 above)
Comment (danielpetisme) [reply generators/server/templates/quarkus/src/test/java/package/TestResources.java.ejs]: Done via the Mock Oidc Test Resource

File: generators/server/templates/quarkus/src/test/resources/application.properties.ejs (hunk 16 above)
Comment (danielpetisme) [reply generators/server/templates/quarkus/src/test/resources/application.properties.ejs]: Done

**Comment by mraible** [comment]
I noticed this error when creating an OAuth app with i18n:

```
? Would you like to enable internationalization support? Yes
? Please choose the native language of the application English
? Please choose additional languages to install Finnish
DEBUG! No blueprint found for blueprint generator-jhipster-quarkus and languages subgenerator: falling back to default generator
DEBUG! Error
Error: Generator jhipster-quarkus:languages isn't registered.
    at module.exports.composeExternalModule (/Users/mraible/generator-jhipster-quarkus/node_modules/generator-jhipster/generators/generator-base.js:1242:19)
    at module.exports.composeBlueprint (/Users/mraible/generator-jhipster-quarkus/node_modules/generator-jhipster/generators/generator-base-private.js:842:49)
    at /Users/mraible/generator-jhipster-quarkus/node_modules/generator-jhipster/generators/generator-base-blueprint.js:126:49
    at Array.forEach (<anonymous>)
    at module.exports.instantiateBlueprints (/Users/mraible/generator-jhipster-quarkus/node_modules/generator-jhipster/generators/generator-base-blueprint.js:118:24)
    at new module.exports (/Users/mraible/generator-jhipster-quarkus/node_modules/generator-jhipster/generators/languages/index.js:86:18)
    at Environment.instantiate (/usr/local/lib/node_modules/generator-jhipster/node_modules/yeoman-environment/lib/environment.js:673:23)
    at instantiate (/Users/mraible/generator-jhipster-quarkus/node_modules/generator-jhipster/node_modules/yeoman-generator/lib/index.js:1196:23)
    at module.exports.composeWith (/Users/mraible/generator-jhipster-quarkus/node_modules/generator-jhipster/node_modules/yeoman-generator/lib/index.js:1227:33)
    at module.exports.composeLanguagesSub (/Users/mraible/generator-jhipster-quarkus/node_modules/generator-jhipster/generators/generator-base.js:1909:23)
DEBUG! Generator is starting

Installing languages: en, fi
```

**Comment by mraible** [comment]
Running `./mvnw` after generating the app results in:

```
2020-12-10 00:17:57,483 ERROR [io.qua.run.Application] (Quarkus Main Thread) Failed to start application (with profile dev): java.net.ConnectException: Connection refused
	at java.base/sun.nio.ch.SocketChannelImpl.checkConnect(Native Method)
	at java.base/sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:779)
	at io.netty.channel.socket.nio.NioSocketChannel.doFinishConnect(NioSocketChannel.java:330)
	at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.finishConnect(AbstractNioChannel.java:334)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:702)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:650)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:576)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

With Spring Security, the app fails to start if Keycloak is not up. So this behavior is consistent, but the message isn't great. With Spring Security, it says the URL that's unavailable.

If I run `jhkeycloakup` to start Keycloak, and then try `./mvnw`, it works. 👍

It doesn't look like there's e2e support for Quarkus yet, so I can't run `npm run e2e` like I can with the main generator. 😕

Trying to go to **Administration** > **Metrics** results in an error. 

image

It seems that many of the "Administration" menu screens don't work.

If I try to generate a Quarkus app with [`21-points.jh`](https://github.com/jhipster/jdl-samples/blob/main/21-points.jh), with the following command:

```
jhipster --blueprint quarkus jdl 21-points.jh
```

But it fails:

```
INFO! Using JHipster version installed globally
INFO! No custom sharedOptions found within blueprint: generator-jhipster-quarkus at /usr/local/lib/node_modules/generator-jhipster-quarkus
INFO! No custom commands found within blueprint: generator-jhipster-quarkus at /usr/local/lib/node_modules/generator-jhipster-quarkus
jdl is not a known command. See 'jhipster --help'.
```

Is JDL supported with the Quarkus blueprint?

**Comment by danielpetisme** [comment]
Thanks @mraible for this feedback.

Regarding the language subgenerator, I think I know the root cause. I've extented the app subgenerator to be able to unit test the `README` generation in a Quarkus context, I guess I have to improve it.

For the error when keycloack isn't up. I do agree the message isn't explicit but I don't consider it as blocking for a first version.

For the admin/metrics part yes it's a known limitation.

For the jdl, I didn't know...

Bottom line, I'll fix the language issue and I consider it good to go.

**Comment by avdev4j** [comment]
> I tried to generate a Quarkus app with 21-points.jh, using the following command:
> 
> jhipster --blueprint quarkus jdl 21-points.jh
> Unfortunately, it fails:
> 
> INFO! Using JHipster version installed globally
> INFO! No custom sharedOptions found within blueprint: generator-jhipster-quarkus at /usr/local/lib/node_modules/generator-jhipster-quarkus
> INFO! No custom commands found within blueprint: generator-jhipster-quarkus at /usr/local/lib/node_modules/generator-jhipster-quarkus
> jdl is not a known command. See 'jhipster --help'.

@mraible 
Could you please retry with the following command:
`jhipster import-jdl 21-points.jh --blueprints quarkus`

I'm not sure `jdl` is correct but fore sure we need to place options (e.g. `--blueprints`) after command.

About admin pages, it's a known bug and we need to fix that before doing a v1. Mainly because of the difference between Spring and Quarkus implementations.

File: generators/server/templates/quarkus/src/main/java/package/web/rest/AccountResource.java.ejs [hunk 17]
```diff
@@ -73,6 +96,119 @@ public class AccountResource {
...
+            return getUserFromAuthentication((JsonWebToken) principal);
+        } else {
+            throw new AccountResourceException("User could not be found");
+        }
+    }
+
+    private static UserVM getUserFromAuthentication(JsonWebToken jwt) {
+        UserVM user = new UserVM();
+        // handle resource server JWT, where sub claim is email and uid is ID
+        if (jwt.getClaim("uid") != null) {
```
Comment (avdev4j) [inline generators/server/templates/quarkus/src/main/java/package/web/rest/AccountResource.java.ejs:137]: This code seems to be the same than in the UserService. Could we have a common method for both?

File: generators/server/templates/quarkus/src/main/java/package/web/rest/UserOauth2Controller.java.ejs [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+import javax.ws.rs.core.UriInfo;
+
+/**
+ * Controller to authenticate users.
+ */
+@Path("/")
+@RequestScoped
+public class UserOauth2Controller {
+    private final Logger log = LoggerFactory.getLogger(UserOauth2Controller.class);
+
```
Comment (avdev4j) [inline generators/server/templates/quarkus/src/main/java/package/web/rest/UserOauth2Controller.java.ejs:42]: remove empty line

File: generators/server/templates/quarkus/src/main/java/package/web/rest/UserResource.java.ejs [hunk 19]
```diff
@@ -149,6 +166,25 @@ public class UserResource {
...
     }
 
+    /**
+     * {@code DELETE /users/:login} : delete the "login" User.
+     *
+     * @param login the login of the user to delete.
+     * @return the {@link Response} with status {@code 204 (NO_CONTENT)}.
+     */
+    @DELETE
+    //    @Path("/{login:" + Constants.LOGIN_REGEX + "}")
```
Comment (avdev4j) [inline generators/server/templates/quarkus/src/main/java/package/web/rest/UserResource.java.ejs:176]: remove?

File: generators/server/templates/quarkus/src/test/java/package/TestResources.java.ejs [hunk 20]
```diff
@@ -23,10 +23,13 @@ import io.quarkus.test.common.QuarkusTestResource;
 import io.quarkus.test.h2.H2DatabaseTestResource;
 <%_ } _%>
 
+<%_ if (databaseType !== 'no') { _%>
```
Comment (avdev4j) [inline generators/server/templates/quarkus/src/test/java/package/TestResources.java.ejs:26]: We should keep sql to avoid problem in the futur with noSql db

File: test/common.spec.js [hunk 21]
```diff
@@ -10,4 +10,30 @@ describe('Subgenerator common of quarkus JHipster blueprint', () => {
             assert.fileContent('README.md', /JHipster Quarkus/);
         });
     });
+
+    // describe('With OAuth2 config', () => {
```
Comment (avdev4j) [inline test/common.spec.js:14]: Remove this part, if this is not necessary to keep it?

File: test/utils/expected-files.js [hunk 22]
```diff
@@ -22,18 +22,17 @@ const expectedFiles = {
...
         common: [
             `${SERVER_MAIN_RES_DIR}application.properties`,
-            `${SERVER_MAIN_RES_DIR}jwt/privateKey.pem`,
-            `${SERVER_MAIN_RES_DIR}META-INF/resources/publicKey.pem`,
             `${SERVER_MAIN_SRC_DIR}com/mycompany/myapp/config/Constants.java`,
             `${SERVER_MAIN_SRC_DIR}com/mycompany/myapp/config/JHipsterProperties.java`,
             `${SERVER_MAIN_SRC_DIR}com/mycompany/myapp/config/JsonbConfiguration.java`,
-            `${SERVER_MAIN_SRC_DIR}com/mycompany/myapp/security/jwt/TokenProvider.java`,
             `${SERVER_MAIN_SRC_DIR}com/mycompany/myapp/web/rest/errors/BadRequestAlertException.java`,
+            // `${SERVER_MAIN_SRC_DIR}com/mycompany/myapp/web/rest/AccountResource.java`,
```
Comment (avdev4j) [inline test/utils/expected-files.js:29]: remove?

File: test/utils/expected-files.js [hunk 23]
```diff
@@ -22,18 +22,17 @@ const expectedFiles = {
...
             `${SERVER_MAIN_SRC_DIR}com/mycompany/myapp/web/util/HeaderUtil.java`,
             `${SERVER_MAIN_SRC_DIR}com/mycompany/myapp/web/util/ResponseUtil.java`,
             `${SERVER_TEST_SRC_DIR}com/mycompany/myapp/ArchTest.java`,
             `${SERVER_TEST_SRC_DIR}com/mycompany/myapp/TestUtil.java`,
             `${SERVER_TEST_SRC_DIR}com/mycompany/myapp/TestResources.java`
+            // `${SERVER_TEST_SRC_DIR}com/mycompany/myapp/web/rest/AccountResourceTest.java`,
```
Comment (avdev4j) [inline test/utils/expected-files.js:35]: remove?

File: generators/server/templates/quarkus/src/main/java/package/web/rest/AccountResource.java.ejs (hunk 17 above)
Comment (danielpetisme) [reply generators/server/templates/quarkus/src/main/java/package/web/rest/AccountResource.java.ejs:137]: Actually I needed to remove code from the `UserService` when oauth2 is choosen I do not generate UserService and everything happens in `AccountResource`

File: generators/server/templates/quarkus/src/test/java/package/TestResources.java.ejs (hunk 20 above)
Comment (danielpetisme) [reply generators/server/templates/quarkus/src/test/java/package/TestResources.java.ejs]: I've updated the condition for `databaseType === 'sql'

File: test/utils/expected-files.js (hunk 22 above)
Comment (danielpetisme) [reply test/utils/expected-files.js]: Yes It was a reminder to create an issue actually :)

**Review by avdev4j — APPROVED** [review]

**Comment by mraible** [comment]
@danielpetisme Thanks for the tip about putting `--blueprints` at the end. This works:

```
jhipster jdl 21-points.jh --blueprints quarkus
```

However, it results in an error (probably because of Elasticsearch).

```
The entity Preferences is being updated.

Git repository initialized.
events.js:292
      throw er; // Unhandled 'error' event
      ^

Error: Copying template quarkus/src/main/java/package/web/rest/EntityResource.java.ejs failed. [ReferenceError: /Users/mraible/generator-jhipster-quarkus/generators/entity-server/templates/quarkus/src/main/java/package/web/rest/EntityResource.java.ejs:60
    58| <%_ if (searchEngine === 'elasticsearch') { _%>
    59|
 >> 60| <%- include('../../_partials/imperative/web/search', this); -%>
    61| <%_ } _%>
    62| }
    63|

/Users/mraible/generator-jhipster-quarkus/generators/entity-server/templates/quarkus/src/main/java/package/_partials/imperative/web/search.ejs:34
    32|     @GET
    33|     @Path("/_search/<%= entityApiUrl %>")
 >> 34| <%- include('../common/search', this); -%>
    35|

/Users/mraible/generator-jhipster-quarkus/generators/entity-server/templates/quarkus/src/main/java/package/_partials/imperative/common/search.ejs:27
    25|         log.debug("REST request to search for a page of <%= entityClassPlural %> for query {}", query);<% if (viaService) { %>
    26|         Page<<%= entityOrDtoClass %>> page = <%= entityInstance %>Service.search(query, pageable);<% } else { %>
 >> 27|         Page<<%= asEntity(entityClass) %>> page = <%= entityInstance %>SearchRepository.search(queryStringQuery(query), pageable);<% } %>
    28|         HttpHeaders headers = PaginationUtil.generatePaginationHttpHeaders(<% if (!reactive) { %>ServletUriComponentsBuilder.fromCurrentRequest()<% } else { %>UriComponentsBuilder.fromHttpRequest(request)<% } %>, page);
    29|         return ResponseEntity.ok().headers(headers).body(<% if (!viaService && dto === 'mapstruct') { %><%= entityToDtoMethodInvocation %>(<% } %>page.getContent()<% if (!viaService && dto === 'mapstruct') { %>)<% } %>);
    30|     <% } -%>

asEntity is not defined]
    at module.exports.error (/Users/mraible/generator-jhipster-quarkus/node_modules/generator-jhipster/generators/generator-base.js:1590:15)
    at /Users/mraible/generator-jhipster-quarkus/node_modules/generator-jhipster/generators/utils.js:256:23
    at tryHandleCache (/Users/mraible/generator-jhipster-quarkus/node_modules/generator-jhipster/node_modules/ejs/lib/ejs.js:275:14)
    at Object.exports.renderFile (/Users/mraible/generator-jhipster-quarkus/node_modules/generator-jhipster/node_modules/ejs/lib/ejs.js:489:10)
    at Object.renderContent (/Users/mraible/generator-jhipster-quarkus/node_modules/generator-jhipster/generators/utils.js:252:9)
    at module.exports.template (/Users/mraible/generator-jhipster-quarkus/node_modules/generator-jhipster/generators/generator-base-private.js:769:23)
    at /Users/mraible/generator-jhipster-quarkus/node_modules/generator-jhipster/generators/generator-base.js:2040:42
    at Array.forEach (<anonymous>)
    at module.exports.writeFilesToDisk (/Users/mraible/generator-jhipster-quarkus/node_modules/generator-jhipster/generators/generator-base.js:2002:45)
    at module.exports.writeServerFiles (/Users/mraible/generator-jhipster-quarkus/generators/entity-server/files.js:343:18)
Emitted 'error' event on Generator instance at:
    at Immediate.<anonymous> (/Users/mraible/generator-jhipster-quarkus/node_modules/generator-jhipster/node_modules/yeoman-generator/lib/index.js:1007:20)
    at processImmediate (internal/timers.js:461:21)
INFO! App: child process exited with code 1
ERROR! JHipster finished with code 1
```

If I remove `searchEngine elasticsearch`, it generates successfully. 

However, if I try to run it, I get an error:

```
FAILURE: Build failed with an exception.

* Where:
Build file '/Users/mraible/quarkus/build.gradle' line: 188

* What went wrong:
A problem occurred evaluating root project 'health-points'.
> Could not get unknown property 'wiremock_version' for object of type org.gradle.api.internal.artifacts.dsl.dependencies.DefaultDependencyHandler.
```

**Comment by mraible** [comment]
I tried generating an app with the [`blog-oauth2.jdl`](https://github.com/jhipster/jdl-samples/blob/main/blog-oauth2.jdl):

```
jhipster jdl blog-oauth2.jdl --blueprints quarkus
```

Unfortunately, this results in a compilation error.

```
[INFO] --- maven-compiler-plugin:3.8.1:compile (default-compile) @ blog ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 30 source files to /Users/mraible/quarkus/target/classes
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR :
[INFO] -------------------------------------------------------------
[ERROR] /Users/mraible/quarkus/src/main/java/com/jhipster/demo/blog/domain/Blog.java:[42,12] cannot find symbol
  symbol:   class User
  location: class com.jhipster.demo.blog.domain.Blog
[INFO] 1 error
[INFO] -------------------------------------------------------------
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  59.470 s
[INFO] Finished at: 2020-12-10T09:09:28-07:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (default-compile) on project blog: Compilation failure
[ERROR] /Users/mraible/quarkus/src/main/java/com/jhipster/demo/blog/domain/Blog.java:[42,12] cannot find symbol
[ERROR]   symbol:   class User
[ERROR]   location: class com.jhipster.demo.blog.domain.Blog
[ERROR]
[ERROR] -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
```

**Comment by mraible** [comment]
I'll create issues for these bugs.
